Document OpenID Connect (#2830)
owncloud/docs: modules/admin_manual/pages/configuration/user/oidc/index.adoc

:openid-connect-app: https://marketplace.owncloud.com/apps/openidconnect
:konnect: https://github.com/Kopano-dev/konnect
:konnect-docs: https://github.com/Kopano-dev/konnect#running-konnect
:konnect-webserver: https://documentation.kopano.io/kopanocore_administrator_manual/configure_kc_components.html#configure-a-webserver-for-konnect

== Introduction

{openid-connect-url}[OpenID Connect] is an open standard for single sign-on and identity and access management. With ownCloud it can be used for user authentication and client authorization against an external Identity Provider.

**Benefits of using ownCloud with OpenID Connect**

- Increased security by shifting user authentication to an external Identity Provider
- Seamless integration into single sign-on (SSO) environments as well as with third party products
- Centralized client management within the Identity Provider
- Enterprise-grade security through the use of authentication security features (e.g., multi-factor authentication) and policies (e.g., automatic token expiration on certain conditions) provided by Identity Providers

== Prerequisites
- OpenID Connect requires an external Identity Provider

> - A distributed memcache setup is required to operate this app - such as Redis or Memcached. Please follow ref:configuration/server/caching_configuration.adoc[the documentation] on how to set up caching.

Reviewer: ```suggestion
- A distributed memcache setup is required to operate this app - such as Redis or Memcached. Please follow xref:configuration/server/caching_configuration.adoc[the documentation] on how to set up caching.
```

Reviewer 2: Fixed in PR #2837 and backport fixed in #2838

== Introduction

{openid-connect-url}[OpenID Connect] is an open standard for single sign-on and identity and access management. With ownCloud it can be used for user authentication and client authorization against an external Identity Provider.

**Benefits of using ownCloud with OpenID Connect**

- Increased security by shifting user authentication to an external Identity Provider
- Seamless integration into single sign-on (SSO) environments as well as with third party products
- Centralized client management within the Identity Provider
- Enterprise-grade security through the use of authentication security features (e.g., multi-factor authentication) and policies (e.g., automatic token expiration on certain conditions) provided by Identity Providers

== Prerequisites
- OpenID Connect requires an external Identity Provider

> - OpenID Connect requires ownCloud Server to be connected with an https://doc.owncloud.com/server/latest/admin_manual/configuration/user/user_auth_ldap.html[LDAP/AD] (local DB users can't log in via OpenID Connect without further measures)

Reviewer 3: just wrong

- Kopano Konnect available as `https://idp.example.com`

=== Set up and configure the Identity Provider

In order to get your Identity Provider running and ready to be used with ownCloud, you have to obtain Kopano Konnect and run it with a set of configuration values which can be provided as environment variables. See the below sections for details.

Specifically, you have to
- provide basic configuration
- connect Kopano Konnect to LDAP/AD
- set up a reverse proxy to expose the routes required to connect to Kopano Konnect. You'll find instructions in the https://documentation.kopano.io/kopanocore_administrator_manual/configure_kc_components.html#configure-a-webserver-for-konnect[Kopano Konnect documentation].

TIP: Kopano Konnect can be set up via Docker. Images are available on Docker Hub (`kopano/kopano_konnect`).
TIP: For general information on Kopano Konnect, please refer to the https://documentation.kopano.io/kopanocore_administrator_manual/configure_kc_components.html#configure-kopano-konnect[Kopano Documentation]. The configuration below is an example to make Kopano Konnect work with ownCloud Server.

==== Basic configuration

> To run Kopano Konnect, some basic configuration is necessary. Please replace the below environment variables with the paths and values suited for your environment.

Reviewer 3: Personally I'd not document how to setup kopano (or any other 3rd party component), because there is more then enough documentation upstream (more then we can write) and things change over time, which means that our information is outdated sooner or later.

```
- FQDN=idp.example.com

> - allow_client_guests=yes

Reviewer 3: guests - really?

- allow_dynamic_client_registration=yes

> - eckey=/kopano/ssl/meet-kwmserver.pem

Reviewer 3: meet?

- ecparam=/kopano/ssl/ecparam.pem
- encryption_secret_key=/kopano/ssl/konnectd-encryption.key
- identifier_registration_conf=/kopano/ssl/konnectd-identifier-registration.yaml

> - identifier_scopes_conf=/etc/kopano/konnectd-identifier-scopes.yaml

Reviewer 3: and here is starts - you are not describing how the key material is created

> ```
> - FQDN=idp.example.com
> - allow_client_guests=yes
> - allow_dynamic_client_registration=yes
> - eckey=/kopano/ssl/meet-kwmserver.pem
> - ecparam=/kopano/ssl/ecparam.pem
> - encryption_secret_key=/kopano/ssl/konnectd-encryption.key
> - identifier_registration_conf=/kopano/ssl/konnectd-identifier-registration.yaml
> - identifier_scopes_conf=/etc/kopano/konnectd-identifier-scopes.yaml
> - signing_private_key=/kopano/ssl/konnectd-tokens-signing-key.pem
> - validation_keys_path=/kopano/ssl/konnectkeys
> ```

Reviewer 4: Nicer code block:

```suggestion
[source,yaml]
----
- FQDN=idp.example.com
- allow_client_guests=yes
- allow_dynamic_client_registration=yes
- eckey=/kopano/ssl/meet-kwmserver.pem
- ecparam=/kopano/ssl/ecparam.pem
- encryption_secret_key=/kopano/ssl/konnectd-encryption.key
- identifier_registration_conf=/kopano/ssl/konnectd-identifier-registration.yaml
- identifier_scopes_conf=/etc/kopano/konnectd-identifier-scopes.yaml
- signing_private_key=/kopano/ssl/konnectd-tokens-signing-key.pem
- validation_keys_path=/kopano/ssl/konnectkeys
----
```

==== Connect Kopano Konnect to LDAP/AD

> To allow Kopano Konnect to authenticate ownCloud users it is necessary to connect it to the same LDAP/AD server which is used by the ownCloud Server. Below you find the environment variables needed to connect Kopano Konnect to an LDAP/AD server. Please replace them with the values suited for your environment (e.g., from the ownCloud LDAP configuration).

Reviewer 3: you describe on how to setup kopano but not the ldap.
especially the config below highly depends on the ldap setup.

this is inconsistent and will result in not working examples -> bad quality and bad ux - sorry

> ```
> - KONNECT_BACKEND=ldap
> - LDAP_BASEDN=dc=owncloud,dc=com
> - LDAP_BINDDN=cn=admin,dc=owncloud,dc=com
> - LDAP_BINDPW=password
> - LDAP_EMAIL_ATTRIBUTE=mail
> - LDAP_FILTER=(objectClass=organizationalPerson)
> - LDAP_LOGIN_ATTRIBUTE=uid
> - LDAP_NAME_ATTRIBUTE=cn
> - LDAP_SCOPE=sub
> - LDAP_URI=ldap://<ldap-address>:389
> - LDAP_UUID_ATTRIBUTE_TYPE=text
> - LDAP_UUID_ATTRIBUTE=uidNumber
> ```

Reviewer 4: Nicer code block:

```suggestion
[source,yaml]
----
- KONNECT_BACKEND=ldap
- LDAP_BASEDN=dc=owncloud,dc=com
- LDAP_BINDDN=cn=admin,dc=owncloud,dc=com
- LDAP_BINDPW=password
- LDAP_EMAIL_ATTRIBUTE=mail
- LDAP_FILTER=(objectClass=organizationalPerson)
- LDAP_LOGIN_ATTRIBUTE=uid
- LDAP_NAME_ATTRIBUTE=cn
- LDAP_SCOPE=sub
- LDAP_URI=ldap://<ldap-address>:389
- LDAP_UUID_ATTRIBUTE_TYPE=text
- LDAP_UUID_ATTRIBUTE=uidNumber
----
```

==== Register ownCloud Clients

To allow the ownCloud Clients (Web/Desktop/Android/iOS) to interact with Kopano Konnect, you have to register them as clients using Konnect's `identifier-registration.yaml`.
Below you find an exemplary `identifier-registration.yaml`. It is based on the default values for the regular ownCloud Clients and works with the configuration of ownCloud Server provided in the sections below. Other environments might require a different set of values. 

> TIP: When registering ownCloud as and OpenID Client, use `https://cloud.example.net/index.php/apps/openidconnect/redirect` as the redirect URL. In case {openid-connect-frontchannel-logout-url}[OpenID Connect Front-Channel Logout 1.0] is supported, please enter `https://cloud.example.net/index.php/apps/openidconnect/logout` as the logout URL within the client registration of the OpenID Provider.

Reviewer 3: afaik kopano is not supporting front-channel-logout (at least it did not months ago) so it is pointless to talk about it here.
please verify with upstream

**Parameters**

> - `id` - the client ID (has to match the respective ID provided by the client)
> - `secret` - the client secret (has to match the respective secret provided by the client)
> - `application_type` - `web` client or `native` application
> - `name` - Display name of the client for the Kopano Konnect login
> - `insecure` - ???
> - `trusted` - if set to `true`, the consent page after successful authentication is skipped and automatically accepted for the user
> - `redirect_uris` - the URLs to redirect to after successful authentication

Reviewer 4: Nice parameter table:

```suggestion

[cols=2,options=header]
|===
| Parameters
| Description

| `id`
| the client ID (has to match the respective ID provided by the client)

| `secret`
| the client secret (has to match the respective secret provided by the client)

| `application_type`
| `web` client or `native` application

| `name` 
| Display name of the client for the Kopano Konnect login

| `insecure`
| ???

| `trusted`
| if set to `true`, the consent page after successful authentication is skipped and automatically accepted for the user

| `redirect_uris` 
|  the URLs to redirect to after successful authentication
|===
```

> ```
> # OpenID Connect client registry
> clients:
>   - id: ownCloud 
>     secret: ownCloud
>     application_type: web
>     insecure: true (???)
> 
>   - id: xdXOt13JKxym1B1QcEncf2XDkLAexMBFwiT9j6EfhhHFJhs2KM9jbjTmf8JBXE69
>     secret: UBntmLjC2yYCeHwsyj73Uwo9TAaecAetRwMw0xYcvNL9yRdLSUi0hUAHfvCHFeFh
>     name: ownCloud desktop client
>     application_type: native
>     insecure: true
>     trusted: true (can both `insecure` and `trusted` work in parallel?)
> 
>   - id: e4rAsNUSIUs0lF4nbv9FmCeUkTlV9GdgTLDH1b5uie7syb90SzEVrbN7HIpmWJeD
>     secret: dInFYGV33xKzhbRmpqQltYNdfLdJIfJ9L5ISoKhNoT9qZftpdWSP71VrpGR9pmoD
>     application_type: native
>     name: ownCloud android app
>     redirect_uris:
>       - oc://android.owncloud.com
> 
>   - id: mxd5OQDk6es5LzOzRvidJNfXLUZS2oN3oUFeXPP8LpPrhx3UroJFduGEYIBOxkY1
>     secret: KFeFWWEZO9TkisIQzR3fo7hfiMXlOpaqP8CFuTbSHzV1TUuGECglPxpiVKJfOXIx
>     name: ownCloud iOS app
>     application_type: native
>     redirect_uris:
>       - oc://ios.owncloud.com
>       - oc.ios://ios.owncloud.com
> ```

Reviewer 4: Alternative code block syntax:

```suggestion
[source,yaml]
----
# OpenID Connect client registry
clients:
  - id: ownCloud 
    secret: ownCloud
    application_type: web
    insecure: true (???)

  - id: xdXOt13JKxym1B1QcEncf2XDkLAexMBFwiT9j6EfhhHFJhs2KM9jbjTmf8JBXE69
    secret: UBntmLjC2yYCeHwsyj73Uwo9TAaecAetRwMw0xYcvNL9yRdLSUi0hUAHfvCHFeFh
    name: ownCloud desktop client
    application_type: native
    insecure: true
    trusted: true (can both `insecure` and `trusted` work in parallel?)

  - id: e4rAsNUSIUs0lF4nbv9FmCeUkTlV9GdgTLDH1b5uie7syb90SzEVrbN7HIpmWJeD
    secret: dInFYGV33xKzhbRmpqQltYNdfLdJIfJ9L5ISoKhNoT9qZftpdWSP71VrpGR9pmoD
    application_type: native
    name: ownCloud android app
    redirect_uris:
      - oc://android.owncloud.com

  - id: mxd5OQDk6es5LzOzRvidJNfXLUZS2oN3oUFeXPP8LpPrhx3UroJFduGEYIBOxkY1
    secret: KFeFWWEZO9TkisIQzR3fo7hfiMXlOpaqP8CFuTbSHzV1TUuGECglPxpiVKJfOXIx
    name: ownCloud iOS app
    application_type: native
    redirect_uris:
      - oc://ios.owncloud.com
      - oc.ios://ios.owncloud.com
----
```

=== Configure ownCloud Server

To set up ownCloud Server to work with OpenID Connect, you have to
1. Install the OpenID Connect App
2. Configure config.php
3. Set up service discovery

TIP: It is recommended to first figure out all configuration on a test system and to bring it to the production system once it's proven to work. Enabling the OpenID Connect App on the production system should be the last step in this process as it will then advertise OpenID Connect to all clients.

==== List of OpenID Connect config.php parameters

Nicht besser hier?
https://doc.owncloud.com/server/admin_manual/configuration/server/config_apps_sample_php_parameters.html

Gibts schon was: https://doc.owncloud.com/server/admin_manual/configuration/server/config_apps_sample_php_parameters.html#openid-connectoidc-configuration

Below you find a list of all `config.php` parameters available to configure OpenID Connect on ownCloud Server.

> - `provider-url` - the URL of the Identity Provider (for some Identity Providers like Keycloak or Azure AD this parameter holds more then just a domain but also a path)
> - `client-id` - the client ID for the ownCloud Web interface as specified in the Identity Provider client registration
> - `client-secret` - the client secret for the ownCloud Web interface as specified in the Identity Provider client registration
> - `mode` - the mode to search for users in ownCloud - possible values are `userid` or `email`
> - `search-attribute` - the attribute which is taken from the access token JWT or user info endpoint to identify a user (?? names as claim as well ??)
> - `loginButtonName` - by default the OpenID Connect App will add a button on the login page that will redirect the user to the Identity Provider and allow authentication via OIDC. This parameter allows to modify the button text.
> - `autoRedirectOnLoginPage` - if set to `true`, the ownCloud login page will redirect directly to the Identity Provider login without requiring the user to click a button.
> - `redirect-url` - the full URL under which the ownCloud OpenID Connect redirect url is reachable (only needed in special setups)
> - `insecure` - if set to `true`, no ssl verification will take place when connecting to the Identity Provider (DON'T use this in production)
> - `scopes` - depending on the setup and needs of the Identity Provider, the list of required scopes is entered here
> - `provider-params` - additional configuration depending on the Identity Provider can be entered here (usually only necessary if the Identity Provider does not support service discovery)
> - `auth-params` - additional parameters which are sent to the Identity Provider during the auth requests
> - `use-token-introspection-endpoint` - if set to `true`, the token introspection endpoint is used to verify a given access token (only needed if the access token is not a JWT)
> - `token-introspection-endpoint-client-id` - the client id to be used with the token introspection endpoint
> - `token-introspection-endpoint-client-secret` - the client secret to be used with the token introspection endpoint
> - `post_logout_redirect_uri` - a given URL where the Identity Provider should redirect to after logout
> - `allowed-user-backends` - use this parameter to limit the users which are allowed to login to a specific user backend (e.g., LDAP) if you want to prevent local db users to log in without OpenID Connect => VERIFY + EXAMPLE
> - `use-access-token-payload-for-user-info` - if set to `true`, any user information will be read from the access token. If set to `false`, the userinfo endpoint is used (requires version >= 1.1.0)
> 
> => needs table!

Reviewer 4: Nice parameter table:

```suggestion
[cols=2,options=header]
|===
| Parameters
| Description

| `provider-url`
| the URL of the Identity Provider (for some Identity Providers like Keycloak or Azure AD this parameter holds more then just a domain but also a path)

| `client-id`
| the client ID for the ownCloud Web interface as specified in the Identity Provider client registration

| `client-secret`
| the client secret for the ownCloud Web interface as specified in the Identity Provider client registration

| `mode`
| the mode to search for users in ownCloud - possible values are `userid` or `email`

| `search-attribute`
| the attribute which is taken from the access token JWT or user info endpoint to identify a user (?? names as claim as well ??)

| `loginButtonName`
| by default the OpenID Connect App will add a button on the login page that will redirect the user to the Identity Provider and allow authentication via OIDC. This parameter allows to modify the button text.

| `autoRedirectOnLoginPage`
| if set to `true`, the ownCloud login page will redirect directly to the Identity Provider login without requiring the user to click a button.

| `redirect-url`
| the full URL under which the ownCloud OpenID Connect redirect url is reachable (only needed in special setups)

| `insecure`
| if set to `true`, no ssl verification will take place when connecting to the Identity Provider (DON'T use this in production)

| `scopes`
| depending on the setup and needs of the Identity Provider, the list of required scopes is entered here

| `provider-params`
| additional configuration depending on the Identity Provider can be entered here (usually only necessary if the Identity Provider does not support service discovery)

| `auth-params`
| additional parameters which are sent to the Identity Provider during the auth requests

| `use-token-introspection-endpoint`
| if set to `true`, the token introspection endpoint is used to verify a given access token (only needed if the access token is not a JWT)

| `token-introspection-endpoint-client-id`
| the client id to be used with the token introspection endpoint

| `token-introspection-endpoint-client-secret`
| the client secret to be used with the token introspection endpoint

| `post_logout_redirect_uri`
| a given URL where the Identity Provider should redirect to after logout

| `allowed-user-backends`
| use this parameter to limit the users which are allowed to login to a specific user backend (e.g., LDAP) if you want to prevent local db users to log in without OpenID Connect => VERIFY + EXAMPLE

| `use-access-token-payload-for-user-info`
| if set to `true`, any user information will be read from the access token. If set to `false`, the userinfo endpoint is used (requires version >= 1.1.0)
|===
```

==== Example configuration for Kopano Konnect

> Below you find an example snippet that can be added to `config.php`. It is based on the above configuration of the Identity Provider (Kopano Konnect). Other environments might require a different set of parameters and values.

Reviewer 2: ```suggestion
An example snippet that can be added to `config.php` is shown below. It is based on the above configuration of the Identity Provider (Kopano Konnect). Other environments might require a different set of parameters and values.
```

```
'openid-connect' =>
array (
  'provider-url' => 'https://idp.example.com',
  'client-id' => 'ownCloud',

>   'client-secret' => 'ownCloud',

Reviewer 3: having the identical value for id and secret is really too simplistic. 
what about using some more descriptive values like
'owncloud-server-client-id' and 'owncloud-server-client-secret'
(same for other clients ...)

  'loginButtonName' => 'Kopano',
  'autoRedirectOnLoginPage' => false,

>   'redirect-url' => 'https://cloud.example.com/index.php/apps/openidconnect/redirect',

Reviewer 3: Not really needed - this is the default url the server can generate itself. This parameter has only to be set in special setups ....

  'mode' => 'userid',

>   'search-attribute' => 'preferred_username',

Reviewer 3: (I did not integrate with kopano for a long time - is this a working config? Please verify!)

>   'use-token-introspection-endpoint' => false,

Reviewer 3: this is the default value - no need to set it

> ```
> 'openid-connect' =>
> array (
>   'provider-url' => 'https://idp.example.com',
>   'client-id' => 'ownCloud',
>   'client-secret' => 'ownCloud',
>   'loginButtonName' => 'Kopano',
>   'autoRedirectOnLoginPage' => false,
>   'redirect-url' => 'https://cloud.example.com/index.php/apps/openidconnect/redirect',
>   'mode' => 'userid',
>   'search-attribute' => 'preferred_username',
>   'use-token-introspection-endpoint' => false,
> ),
> ```

Reviewer 4: Alternative code block syntax:

```suggestion
[source,php]
----
'openid-connect' =>
array (
  'provider-url' => 'https://idp.example.com',
  'client-id' => 'ownCloud',
  'client-secret' => 'ownCloud',
  'loginButtonName' => 'Kopano',
  'autoRedirectOnLoginPage' => false,
  'redirect-url' => 'https://cloud.example.com/index.php/apps/openidconnect/redirect',
  'mode' => 'userid',
  'search-attribute' => 'preferred_username',
  'use-token-introspection-endpoint' => false,
),
----
```

==== Set up service discovery

In order to allow the ownCloud Clients (Desktop/Android/iOS) to make use of OpenID Connect, ownCloud Server needs to provide service discovery information under the static path `https://cloud.example.com/.well-known/openid-configuration`.
When enabled, the OpenID Connect App provides the service discovery information on the endpoint `https://cloud.example.com/index.php/apps/openidconnect/config`.

> To make the endpoint available under the static service discovery path, it is recommended to put a `RewriteRule` in place using `.htaccess` (the Apache modules `proxy` and `proxy_http` have to be enabled):

Reviewer 3: For the sake of completness - this rewrite rule can be setup in the apache config as well ....

> `RewriteRule ^\.well-known/openid-configuration /index.php/apps/openidconnect/config [P]`

Reviewer 3: we should change this in core/.htaccess
